src/main/sensors/gyroanalyse.c

```diff
@@ -46,58 +46,85 @@
 // Eg [0,31), [31,62), [62, 93) etc
 
 #define FFT_BIN_COUNT             (FFT_WINDOW_SIZE / 2)
-// compare 1 + this offset FFT bins for peak, ie if 1 start 2.5 * 41.655 or about 104Hz
-#define FFT_BIN_OFFSET            1
-// analyse up to 666Hz, 16 bins each 41.625 Hz wide
+// for gyro loop >= 4KHz, analyse up to 666Hz, 16 bins each 41.625 Hz wide
 #define FFT_SAMPLING_RATE_HZ      1333
-// centre frequency of bandpass that constrains input to FFT
-#define FFT_BPF_HZ                (FFT_SAMPLING_RATE_HZ / 4)
-// Hz per bin
-#define FFT_RESOLUTION            ((float)FFT_SAMPLING_RATE_HZ / FFT_WINDOW_SIZE)
 // following bin must be at least 2 times previous to indicate start of peak
 #define FFT_MIN_BIN_RISE          2
 
 // lowpass frequency for smoothing notch centre point
 #define DYN_NOTCH_SMOOTH_FREQ_HZ  60
 // notch centre point will not go below this, must be greater than cutoff, mid of bottom bin
 #define DYN_NOTCH_MIN_CENTRE_HZ   125
-// maximum notch centre frequency limited by Nyquist
-#define DYN_NOTCH_MAX_CENTRE_HZ   (FFT_SAMPLING_RATE_HZ / 2)
 // lowest allowed notch cutoff frequency
 #define DYN_NOTCH_MIN_CUTOFF_HZ   105
 // we need 4 steps for each axis
 #define DYN_NOTCH_CALC_TICKS      (XYZ_AXIS_COUNT * 4)
 
+static bool FAST_RAM_ZERO_INIT gyroAnalyseInitialized;
+static uint16_t FAST_RAM_ZERO_INIT fftSamplingRateHz;
+// centre frequency of bandpass that constrains input to FFT
+static uint16_t FAST_RAM_ZERO_INIT fftBpfHz;
+// Hz per bin
+static float FAST_RAM_ZERO_INIT    fftResolution;
+// maximum notch centre frequency limited by Nyquist
+static uint16_t FAST_RAM_ZERO_INIT dynNotchMaxCentreHz;
+static uint8_t  FAST_RAM_ZERO_INIT fftBinOffset;
+
 // Hanning window, see https://en.wikipedia.org/wiki/Window_function#Hann_.28Hanning.29_window
 static FAST_RAM_ZERO_INIT float hanningWindow[FFT_WINDOW_SIZE];
 static FAST_RAM_ZERO_INIT float dynamicNotchCutoff;
 
-void gyroDataAnalyseInit(void)
+// Making this NOINLINE saves a few bytes to help F3's compile
+// It's only called once at initialization so no performance concerns
+NOINLINE void gyroDataAnalyseInit(uint32_t targetLooptimeUs)
 {
+    const int gyroLoopRateHz = lrintf((1 / (float)targetLooptimeUs) * 1e6);
+
+    // If we get at least 3 samples then use the default FFT sample frequency
+    // otherwise we need to calculate a FFT sample frequency to ensure we get 3 samples (gyro loops < 4K)
```

Comment:
Why do you want at least 3 samples? The filter should work fine down to 1kHz gyro sampling rate by using raw data instead of accumulated. At least this keeps the usable centerFreq up to 500Hz.

Reply:
I need to pull @ctzsnooze into this conversation.

The original problem we found was related to changing the FFT sampling to 1333hz.  When running at 2KHz the `centerFreq` calc was so chaotic from lack of samples that it ended up causing the center lowpass result to overshoot beyond the MIN/MAX and sometimes go negative - leading to an underflow when loaded into a `uint16_t`.  So part of this PR is to add an extra `constrain` on the centerFreq lowpass to prevent.

So with that fixed that lead to looking at the results of the centerFreq calc and the opinion was that it wasn't working well without more samples.  I don't know if that's a result of the other changes to how the peak bin/center frequency is selected or a previous problem.  Adjusting the FFT sampling to ensure 3 samples seemed to product more consistent and better results - but of course at the loss of width at lower loop times.

Reply:
Well you can't derive 1333hz from 2000 hz, as it would be 2000/1333 = 1.5 samples, which proably explains some erratic behaviour.
But I think a better approach would be to set FFT sampling rate back to 1000Hz instead of 1333Hz for 2kHz and 1kHz sampling and use 2 / 1 samples instead. This way you keep the resolution up.

Reply:
Sorry, I forgot to mention that we tried setting the FFT sampling back to 1000 with 2K and still the results were not good.  That drove the reason to use 3 samples.

Frankly I don't think we're worried about supporting < 2K anymore.  The use case is basically non-existent now that F1's are no longer supported.

Reply:
I will look at it more in 2k with 2/1 at 1k